alica_tracing/include/tracing/AlicaTrace.h

```diff
@@ -0,0 +1,197 @@
+    {
+        if (parent) {
+            std::stringstream ss(parent.value());
+            jaegertracing::SpanContext parentCtx = jaegertracing::SpanContext::fromStream(ss);
+            _rawTrace = opentracing::Tracer::Global()->StartSpan(opName, {opentracing::ChildOf(&parentCtx)});
+        } else {
+            _rawTrace = opentracing::Tracer::Global()->StartSpan(opName);
+            if (opName == "MasterPlan") {
+                _rawTrace->Finish();
+            }
+        }
+    }
+
+    Trace() {}
+
+    Trace(Trace&& other) = default;
+
+    Trace& operator=(Trace&& other) = default;
+
+    // This class can either be used raw, inherited or can be used to compose module level trace classes
+    // Destruction of this instance will mark the end time of operation, and will be reported to tracing collectors
+    virtual ~Trace()
+    {
+        if (!_rawTrace) {
+            return;
+        }
+
+        _rawTrace->Finish();
+    }
+
+    // Tags are the key value pair you can search in tracing ui
+    // If there is a pre-existing tag set for `key`, it is overwritten.
+    // Tag values can be numeric types, strings, or bools.
+    void setTag(const std::string& key, const std::string& value)
+    {
+        if (!_rawTrace) {
+            return;
+        }
+
+        _rawTrace->SetTag(key, value);
+    }
+
+    // setLog is a timestamped way to record key:value logging data
+    // about a trace. Here's an example:
+    //
+    //    trace.Log({
+    //        {"progress", "checkpoint 4 cleared"},
+    //        {"wait.time", "waited 10 sec for agent 2"}});
+    // BE CAREFUL about what you want to log,
+    // These are supposed to be micro logs to be carried over the network.
+    void setLog(std::pair<std::string, std::string> logEntry)
+    {
+        if (!_rawTrace) {
+            return;
+        }
+
+        _rawTrace->Log({logEntry});
+    }
+
+    // When operation being traced by this instance fail, call this api
+    // Error traces are highlighted in tracing ui.
+    void markError(const std::string& description)
+    {
+        if (!_rawTrace) {
+            return;
+        }
+
+        _rawTrace->SetTag("error", true);
+        _rawTrace->SetTag("error.description", description);
+    }
+
+    // Get the context of this trace to propogate across process boundary
+    std::string context() const
+    {
+        if (!_rawTrace) {
+            return "";
+        }
+
+        const auto& jctx = dynamic_cast<const jaegertracing::SpanContext&>(_rawTrace->context());
+        std::stringstream ss;
+        jctx.print(ss);
+        return ss.str();
+    }
+
+private:
+    // This could be intentionally empty in the case when factory isn't initialized.
+    // An app can choose not to initialize the TraceFactory and
+    // as a result, traces should become no-op.
+    RawTracePtr _rawTrace;
+};
+
+using TracePtr = std::unique_ptr<Trace>;
+
+template <class TraceType>
```

Comment:
Don't see any use in making this a template because this factory can't create traces of any other type.